### PR TITLE
?include_revisions=:bool parameter 

### LINKED ISSUE #198: [13] Add "include_revisions" parameter
This parameter will include the 'parents' field with an aduit list of the previous revisions of an object to any request made (to just an endpoint, or to a specific object of an endpoint). If this flag is not passed just the most recent revision of an object is returned.

### PR DESCRIPTION
**Pull request by ElijahCaine**
resolves #198 when merged

### CODE REVIEW

File: tests/activities.js [hunk 1]
```diff
@@ -889,4 +889,218 @@ module.exports = function(expect, request, baseUrl) {
...
+      const postArg = getPostObject(baseUrl + 'activities/' + activity,
+                      postActivity);
+
+      request.post(postArg, function() {
+        done();
+      });
+    });
+
+    // Tests that `include_revisions=true` includes revisions
+    it('gets activities + revisions when include_revions=true is passed',
```
Comment (LyonesGamer) [inline tests/activities.js]: typo: "include_revions"

File: tests/activities.js [hunk 2]
```diff
@@ -889,4 +889,218 @@ module.exports = function(expect, request, baseUrl) {
...
+      request.get(baseUrl + 'activities/?include_revisions=true',
+      function(err, res, body) {
+        expect(JSON.parse(body)).to.include(withParentsData);
+        expect(JSON.parse(body)).to.not.include(noParentsData);
+        done();
+      });
+    });
+
+    // Tests that `include_revisions` includes revisions
+    it('gets activities + revisions when include_revions is passed',
```
Comment (LyonesGamer) [inline tests/activities.js]: typo: "include_revions"

**Comment by Kennric** [comment]
There's a lot in here - other than typos as mentioned by @LyonesGamer, this looks pretty good, I'll approve when typos are cleaned up.

File: tests/activities.js [hunk 3]
```diff
@@ -889,4 +889,218 @@ module.exports = function(expect, request, baseUrl) {
...
+      request.get(baseUrl + 'activities/' + activity + '?include_revisions',
+      function(err, res, body) {
+        expect(JSON.parse(body)).to.include(withParentsData);
+        expect(JSON.parse(body)).to.not.include(noParentsData);
+        done();
+      });
+    });
+
+    // Tests that include_revisions isn't always set to true
+    it('gets just acitivity when include_revisions=false', function(done) {
```
Comment (athai) [inline tests/activities.js]: acitivity != activity

File: tests/activities.js [hunk 4]
```diff
@@ -889,4 +889,218 @@ module.exports = function(expect, request, baseUrl) {
...
+      request.get(baseUrl + 'activities/' + activity +
+      '?include_revisions=false', function(err, res, body) {
+        expect(JSON.parse(body)).to.include(noParentsData);
+        done();
+      });
+    });
+
+    // Tests that include_revisions defaults to false
+    it('gets just acitivity when include_revisions is not set',
```
Comment (athai) [inline tests/activities.js]: acitivity != activity

**Comment by ElijahCaine** [comment]
@osuosl/devs @Kennric 
Current Status of this PR (in case I'm not able to work for the next few weeks):
* I am implementing a field on the backend (``newest``) which will help in building and maintaining include_revisions.
* I also made all of this include_revisions work and the required changes to the current system one PR instead of spreading it across a few PR's. This will help with merging since a lot of things on develop are changing and I'd rather deal with the conflicts as they arise and when the changes are fresh in our minds.
* I also deprecated the 'id' field from the returned statements since that's a thing we haven't done yet. refs.

It's not done but with the new field it should go a bit smoother.

**Comment by LyonesGamer** [comment]
Will also resolve #222 when merged.

**Comment by ElijahCaine** [comment]
@osuosl/devs and @Kennric @LyonesGamer in particular I'd like to get this reviewed and merged ASAP. If you have any questions please don't hesitate to ask.

There are a few sections of this code I mean to document further but for one reason or another I neglected to do so. I am glad to further document, just point at sections that are unclear.

File: src/projects.js [hunk 5]
```diff
@@ -19,55 +51,81 @@ module.exports = function(app) {
...
+          const slugs = projects.filter(function(p) {
+            return p.uuid === proj.uuid && p.slug;
+          }).map(function(p) {
+            return p.slug;
+          });
 
-        return res.send(projects);
+          return JSON.stringify(constructProject(proj, res, slugs));
+        }).filter(function(proj, index, self) {
+          return index === self.indexOf(proj);
```
Comment (LyonesGamer) [inline src/projects.js]: I'd leave a comment here explaining this.

File: src/activities.js [hunk 6]
```diff
@@ -6,6 +6,33 @@ module.exports = function(app) {
...
+    } else {
+      activity.updated_at = null;
+    }
+    if (activity.deleted_at) {
+      activity.deleted_at = new Date(parseInt(activity.deleted_at, 10))
+      .toISOString().substring(0, 10);
+    } else {
+      activity.deleted_at = null;
+    }
+    delete activity.id;
```
Comment (LyonesGamer) [inline src/activities.js]: You should standardize these to use the same method, either delete fields or manually construct the object.

File: src/projects.js [hunk 7]
```diff
@@ -7,6 +7,38 @@ module.exports = function(app) {
...
 
+  function constructProject(inProject, res, slugs) {
+    if (!inProject) {
+      const err = errors.errorObjectNotFound('project');
+      res.status(err.status);
+      return err;
+    }
+
+    // manually create our project object from inProject.
+    const outProject = {name: inProject.name, uri:
```
Comment (LyonesGamer) [inline src/projects.js]: You should standardize these to use the same method, either delete fields or manually construct the object.

File: src/times.js [hunk 8]
```diff
@@ -7,425 +7,468 @@ module.exports = function(app) {
...
-            }
-          }
-
-          selectedActivities = selectedActivities.filter(function(activity) {
-            return activitiesList.indexOf(activity.slug) !== -1;
-          });
+    time.project = project.sort();
+    time.activities = activities.sort();
+
+    delete time.activity;
```
Comment (LyonesGamer) [inline src/times.js]: You should standardize these to use the same method, either delete fields or manually construct the object.

File: src/times.js [hunk 9]
```diff
@@ -7,425 +7,468 @@ module.exports = function(app) {
...
-          knex('projectslugs').then(function(projectslugs) {
-            if (projectsList && projectsList.length) {
-              const projectIds = projectslugs.filter(function(slug) {
-                // Filter down to only slugs that were requested
-                return projectsList.indexOf(slug.name) !== -1;
-              }).map(function(slug) {
-                return slug.project;
-              });
+          if (projectArr) {
+            // Check that the user we just parsed is in the database
```
Comment (Kennric) [inline src/times.js]: You probably mean 'project', not 'user'

File: src/projects.js [hunk 10]
```diff
@@ -443,53 +489,62 @@ module.exports = function(app) {
...
-                  if (obj.slugs) {
-                    const newSlugs = [];
-
-                    newSlugs.push(trx('projectslugs').del()
-                    .where({project: oldId}));
+            trx('projects').update({newest: false}).where({id: oldId})
+            .then(function() {
+              trx('projects').insert(project).returning('id')
+              .then(function(id) {
+                project.id = id[0];
```
Comment (LyonesGamer) [inline src/projects.js]: We set project.id, only to delete it later. Maybe best to just change this to a local variable.

File: src/times.js [hunk 11]
```diff
@@ -7,425 +7,468 @@ module.exports = function(app) {
...
-                // processing finished. Return if others are also finished
-                usersDone = true;
-                if (activitiesDone && projectsDone) {
-                  return res.send(times);
-                }
-              }).catch(function(error) {
-                const err = errors.errorServerError(error);
-                return res.status(err.status).send(err);
+          if (activityArr) {
+            // Check that the user we just parsed is in the database
```
Comment (Kennric) [inline src/times.js]: user -> activity

File: src/times.js [hunk 12]
```diff
@@ -7,425 +7,468 @@ module.exports = function(app) {
...
+                     helpers.getType(req.query.end[0]) === 'string') {
+            // Set range's end value to the first end given
+            range[1] = req.query.end[0];
+          // Set to undefined to later be set to a sane value
+          } else { range[1] = undefined; }
+
+          const currDate = new Date().getTime();
+          // If the value is undefined, leave it as such
+          if (range[0] === undefined) {
+            // Run a regex test on the times in the query parameter
```
Comment (LyonesGamer) [inline src/times.js]: I think you might be missing a bit of code here.

File: src/times.js [hunk 13]
```diff
@@ -7,425 +7,468 @@ module.exports = function(app) {
...
+            reject(res.status(err.status).send(err));
+          }
 
-            if (!startDate || !startDate.getTime() ||
-            startDate.getTime() > Date.now()) {
+          // Both end and start are specified
+          if (!isNaN(range[0]) && !isNaN(range[1])) {
+            // Test that the times submitted were valid within the range of
+            // possible dates.
+            if (!range[0] || range[0] > Date.now()) {
```
Comment (LyonesGamer) [inline src/times.js]: Didn't you test this on line 118?

File: src/times.js [hunk 14]
```diff
@@ -740,55 +790,61 @@ module.exports = function(app) {
...
-                    /* eslint-disable prefer-const */
-                    for (let activity of activities) {
-                      /* eslint-enable prefer-const */
-                      taInsertion.push({
-                        time: time[0].id,
-                        activity: activity.activity,
+              trx('times').where({uuid: req.params.uuid, newest: true})
+              .update({newest: false}).then(function() {
+                trx('times').insert(time).returning('id').then(function(id) {
+                  time.id = id[0];
```
Comment (LyonesGamer) [inline src/times.js]: Once again, probably make this a local variable since we don't want to return an id.

File: tests/activities.js [hunk 15]
```diff
@@ -1164,4 +1126,232 @@ module.exports = function(expect, request, baseUrl) {
...
+    });
+
+    // Tests that `include_revisions=true` includes revisions
+    it('gets activities + revisions when include_revisions=true is passed',
+    function(done) {
+      getAPIToken().then(function(token) {
+        request.get(baseUrl + 'activities/?include_revisions=true&token=' +
+        token,
+        function(err, res, body) {
+          expect(JSON.parse(body)).to.include(withParentsData);
```
Comment (LyonesGamer) [inline tests/activities.js]: I may be wrong, but I think we want to change these to expect().to.deep.include().

File: tests/activities.js [hunk 16]
```diff
@@ -1164,4 +1126,232 @@ module.exports = function(expect, request, baseUrl) {
...
+        });
+      });
+    });
+
+    // Tests that include_revisions isn't always set to true
+    it('gets just activity when include_revisions=false', function(done) {
+      getAPIToken().then(function(token) {
+        request.get(baseUrl + 'activities/' + activity +
+        '?include_revisions=false&token=' + token, function(err, res, body) {
+          expect(JSON.parse(body)).to.include(noParentsData);
```
Comment (LyonesGamer) [inline tests/activities.js]: And these should probably be ``to.deep.equal()``.

File: src/times.js (hunk 13 above)
Comment (ElijahCaine) [inline src/times.js]: Good catch. I removed the earlier check.

**Comment by Kennric** [comment]
Looking pretty good, please have another review or two. @athai @LyonesGamer 
+1

File: src/projects.js [hunk 17]
```diff
@@ -19,55 +50,88 @@ module.exports = function(app) {
...
-          .toISOString().substring(0, 10);
-        } else {
-          project.deleted_at = null;
-        }
-      }
+        return res.send(projects.filter(function(proj) {
+          return proj.newest;
+        }).map(function(proj) {
+          const slugs = projects.filter(function(p) {
+            return p.uuid === proj.uuid && p.slug;
```
Comment (athai) [inline src/projects.js]: Wouldn't this always return false?

**Comment by LyonesGamer** [comment]
Looks pretty good to me. +1

File: src/times.js [hunk 18]
```diff
@@ -7,425 +7,476 @@ module.exports = function(app) {
...
+      notes: inTime.notes,
+      issue_uri: inTime.issue_uri,
+      uuid: inTime.uuid,
+      revision: inTime.revision,
+      project: project.sort(),
+      activities: activities.sort(),
+      created_at: inTime.created_at,
+      updated_at: inTime.updated_at,
+      deleted_at: inTime.deleted_at,
+      date_worked: inTime.deleted_at,
```
Comment (athai) [inline src/times.js]: Why is date_worked being set to inTime.deleted_at?

File: src/times.js [hunk 19]
```diff
@@ -683,54 +734,61 @@ module.exports = function(app) {
...
+      'times.project as project',
+      'times.notes as notes',
+      'times.issue_uri as issue_uri',
+      'times.date_worked as date_worked',
+      'times.created_at as created_at',
+      'times.updated_at as updated_at',
+      'times.id as id',
+      'times.uuid as uuid',
+      'times.revision as revision',
+      'users.username as owner',
```
Comment (athai) [inline src/times.js]: Remove owner

**Comment by athai** [comment]
+1 Looks good. The api doc should be updated to include an example of include_revisions on the projects endpoint though.

**Comment by LyonesGamer** [comment]
I agree, looks like it's all pretty good. +1

**Comment by athai** [comment]
The change in variable names makes everything much clearer! +1

**Comment by Kennric** [comment]
I'm happy with it.

**Pull request merged**